### ISSUE TITLE
Using wrong case on module should warn

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8556]
Migrated from [rt.perl.org#40131](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=40131) (status was 'rejected')

Searchable as RT40131$

**Comment by p5pRT** [comment]
### From @btilly

#### Created by btilly@gmail.com



Not on this machine \(obviously\)\, but on case insensitive systems like
Windows or OS X Perl generates no warnings or errors if you get the
case of a module wrong\, yet may not do anything useful\.  Consider\, for
instance how much "use Strict;" doesn't do for you\.

I think there should be at least a warning if Perl picks up a module
whose filename's case doesn't match what was requested by use or
require\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=wishlist

Site configuration information for perl v5.8.7:

Configured by Debian Project at Wed Sep 21 13:50:37 UTC 2005.

Summary of my perl5 (revision 5 version 8 subversion 7) configuration:
  Platform:
    osname=linux, osvers=2.6.12, archname=i486-linux-gnu-thread-multi
    uname='linux terranova 2.6.12 #1 smp tue aug 9 18:56:34 utc 2005 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.7 -Dsitearch=/usr/local/lib/perl/5.8.7 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.7 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.0.2 20050808 (prerelease) (Ubuntu 4.0.1-4ubuntu8)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.5.so, so=so, useshrplib=true, libperl=libperl.so.5.8.7
    gnulibc_version='2.3.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.7:
    /etc/perl
    /usr/local/lib/perl/5.8.7
    /usr/local/share/perl/5.8.7
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.7:
    HOME=/home/btilly
    LANG=en_US.UTF-8
    LANGUAGE=en
    LD_LIBRARY_PATH=/usr/local/oracle/lib
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/local/sbin:/sbin:/usr/sbin:/bin:/usr/bin:/usr/bin/X11:/usr/games:/usr/local/oracle/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgarcia

On 09/08/06\, via RT Ben Tilly \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> Not on this machine \(obviously\)\, but on case insensitive systems like
> Windows or OS X Perl generates no warnings or errors if you get the
> case of a module wrong\, yet may not do anything useful\.  Consider\, for
> instance how much "use Strict;" doesn't do for you\.
>
> I think there should be at least a warning if Perl picks up a module
> whose filename's case doesn't match what was requested by use or
> require\.

There used to be such a warning\, but it has been removed; now the
sources contain this comment :

&nbsp;    /\* The "did you use incorrect case?" warning used to be here\.
&nbsp;     \* The problem is that on case\-insensitive filesystems one
&nbsp;     \* might get false positives for "use" \(and "require"\)&#8203;:
&nbsp;     \* "use Strict" or "require CARP" will work\.  This causes
&nbsp;     \* portability problems for the script&#8203;: in case\-strict
&nbsp;     \* filesystems the script will stop working\.
&nbsp;     \*
&nbsp;     \* The "incorrect case" warning checked whether "use Foo"
&nbsp;     \* imported "Foo" to your namespace\, but that is wrong\, too&#8203;:
&nbsp;     \* there is no requirement nor promise in the language that
&nbsp;     \* a Foo\.pm should or would contain anything in package "Foo"\.
&nbsp;     \*
&nbsp;     \* There is very little Configure\-wise that can be done\, either&#8203;:
&nbsp;     \* the case\-sensitivity of the build filesystem of Perl does not
&nbsp;     \* help in guessing the case\-sensitivity of the runtime environment\.
&nbsp;     \*/

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From blgl@hagernas.com

In article \<rt\-3\.5\.HEAD\-22701\-1155094149\-569\.40131\-75\-0@&#8203;perl\.org>\,
 perlbug\-followup@&#8203;perl\.org \(Ben Tilly\) wrote&#8203;:

> I think there should be at least a warning if Perl picks up a module
> whose filename's case doesn't match what was requested by use or
> require\.

You'd need some platform specific code to check whether the
requested name equals the stored name of the file\.  This is
fairly painless on Darwin/Mac OS X \(call getattrlist\(2\) with
ATTR\_CMN\_NAME\); how is it elsewhere?


/Bo Lindbergh

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'stalled'

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Aug 08 23&#8203;:45&#8203;:32 2006\, old rgarciasuarez@&#8203;gmail\.com wrote&#8203;:
> On 09/08/06\, via RT Ben Tilly \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > Not on this machine \(obviously\)\, but on case insensitive systems like
> > Windows or OS X Perl generates no warnings or errors if you get the
> > case of a module wrong\, yet may not do anything useful\.  Consider\, for
> > instance how much "use Strict;" doesn't do for you\.
> >
> > I think there should be at least a warning if Perl picks up a module
> > whose filename's case doesn't match what was requested by use or
> > require\.
> 
> There used to be such a warning\, but it has been removed; now the
> sources contain this comment :
> 
>     /\* The "did you use incorrect case?" warning used to be here\.
>      \* The problem is that on case\-insensitive filesystems one
>      \* might get false positives for "use" \(and "require"\)&#8203;:
>      \* "use Strict" or "require CARP" will work\.  This causes
>      \* portability problems for the script&#8203;: in case\-strict
>      \* filesystems the script will stop working\.
>      \*
>      \* The "incorrect case" warning checked whether "use Foo"
>      \* imported "Foo" to your namespace\, but that is wrong\, too&#8203;:
>      \* there is no requirement nor promise in the language that
>      \* a Foo\.pm should or would contain anything in package "Foo"\.
>      \*
>      \* There is very little Configure\-wise that can be done\, either&#8203;:
>      \* the case\-sensitivity of the build filesystem of Perl does not
>      \* help in guessing the case\-sensitivity of the runtime environment\.
>      \*/
> 

This ticket was in the Stalled queue for six years\.  My sense is that it
is a "wontfix"\.  I have taken the ticket for the purpose of closing it
within seven days unless someone has a serious objection\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'stalled' to 'open'

**Comment by p5pRT** [comment]
### From @bulk88

I dont see a portable way of doing this if you include case insensitive
and not case preserving filing systems like FAT16 on DOS\. The only way
to fix this and be compatible with FAT16 is to do a breaking syntax
change and require all PM files to begin with some kind of comment or
syntax which says the case sensitive name of what that PM file is
supposed to be for the purpose of use and require\. Something similar in
concept to the shebang \(which itself isn't required by Perl today\)\.
Another idea is to keep all PM files in a RDBMS \(sqlite?\) or compressed
archive\, think RPM or MS MSI or Java jar files and all use and requires
that use :&#8203;: style names goto the RDBMS file and not the filing system\.
Basically Perl implements its own filing system for all "&#8203;::" named PMs\.
Both ideas would cause Perl 5 to not be Perl 5 anymore\. The "Perl
implements a filing system for modules" idea would also make it easier
to implement crypto signing and DRM for CPAN modules to prevent rogue
and insecure modules from "untrusted" \(what is an untrusted author? IDK\)
PAUSE authors being installed\. I say wont fix\.

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Oct 07 11&#8203;:28&#8203;:17 2012\, bulk88 wrote&#8203;:
>  I say wont fix\.

Which gives us two thumbs down\.  Closing as scheduled\.

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @rjbs

\* bulk 88 via RT \<perlbug\-followup@&#8203;perl\.org> \[2012\-10\-07T14&#8203;:28&#8203;:18\]
> I dont see a portable way of doing this if you include case insensitive
> and not case preserving filing systems like FAT16 on DOS\.

Do we still support this configuration in general?

That is&#8203;:

1\. specifically\, do we work on a FAT16 file system?

2\. generally\, are there other such filesystems on which we'd like to work?

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From vadim.konovalov@alcatel-lucent.com

> From&#8203;: Ricardo Signes \[mailto&#8203;:perl\.p5p@&#8203;rjbs\.manxome\.org\] 
> \* bulk 88 via RT \<perlbug\-followup@&#8203;perl\.org> \[2012\-10\-07T14&#8203;:28&#8203;:18\]
> > I dont see a portable way of doing this if you include case 
> insensitive
> > and not case preserving filing systems like FAT16 on DOS\.
> 
> Do we still support this configuration in general?
> 
> That is&#8203;:
> 
> 1\. specifically\, do we work on a FAT16 file system?

IMO this FAT16 thing is no different compared to FAT32 and even to NTFS \- 
these are case insensitive\, case forgiving file systems\,

Even this is not refering to 8\.3 restrictions&#8203;: FAT16 is able to deal with long 
names\. \(so DOS perl could suffer some difficulties\, but those are not
related to the ticket\)

I also say "wont fix" and close the ticket :\)

> 
> 2\. generally\, are there other such filesystems on which we'd 
> like to work?

y

**Comment by p5pRT** [comment]
### From @demerphq

On 18 October 2012 11&#8203;:05\, Konovalov\, Vadim \(Vadim\)\*\* CTR \*\*
\<vadim\.konovalov@&#8203;alcatel\-lucent\.com> wrote&#8203;:
>> From&#8203;: Ricardo Signes \[mailto&#8203;:perl\.p5p@&#8203;rjbs\.manxome\.org\]
>> \* bulk 88 via RT \<perlbug\-followup@&#8203;perl\.org> \[2012\-10\-07T14&#8203;:28&#8203;:18\]
>> > I dont see a portable way of doing this if you include case
>> insensitive
>> > and not case preserving filing systems like FAT16 on DOS\.
>>
>> Do we still support this configuration in general?
>>
>> That is&#8203;:
>>
>> 1\. specifically\, do we work on a FAT16 file system?
>
> IMO this FAT16 thing is no different compared to FAT32 and even to NTFS \-
> these are case insensitive\, case forgiving file systems\,

What you just said makes no sense\. This thread is about detecting
misloads due to case insensitive file systems\.

> Even this is not refering to 8\.3 restrictions&#8203;: FAT16 is able to deal with long
> names\. \(so DOS perl could suffer some difficulties\, but those are not
> related to the ticket\)

I dont understand what this has to do with the original ticket?

Yves
\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From vadim.konovalov@alcatel-lucent.com

> From&#8203;: demerphq \[mailto&#8203;:demerphq@&#8203;gmail\.com\] 
> On 18 October 2012 11&#8203;:05\, Konovalov\, Vadim wrote&#8203;:
> >> From&#8203;: Ricardo Signes \[mailto&#8203;:perl\.p5p@&#8203;rjbs\.manxome\.org\]
> >> \* bulk 88 via RT \<perlbug\-followup@&#8203;perl\.org> \[2012\-10\-07T14&#8203;:28&#8203;:18\]
> >> > I dont see a portable way of doing this if you include case
> >> insensitive
> >> > and not case preserving filing systems like FAT16 on DOS\.
> >>
> >> Do we still support this configuration in general?
> >>
> >> That is&#8203;:
> >>
> >> 1\. specifically\, do we work on a FAT16 file system?
> >
> > IMO this FAT16 thing is no different compared to FAT32 and 
> even to NTFS \-
> > these are case insensitive\, case forgiving file systems\,
> 
> What you just said makes no sense\. This thread is about detecting
> misloads due to case insensitive file systems\.

ok\, 

my message was an answer to RJBS's question about FAT16\,

so I pointed that this "FAT16" is exactly the same as FAT32 and NTFS\,
WRT the problem described in ticket\,

and also those those two FSs are quite current\, i\.e\. not obsolete\.

> 
> > Even this is not refering to 8\.3 restrictions&#8203;: FAT16 is 
> able to deal with long
> > names\. \(so DOS perl could suffer some difficulties\, but 
> those are not
> > related to the ticket\)
> 
> I dont understand what this has to do with the original ticket?

maybe I misunderstood the point \- but 1\) original ticket was not 
refering to DOS \(which could have 8\.3 restriction on file names\)
and 2\) was about things like 

use Strict;

Somehow\, DOS appeared later in discussions\, and this 
misleading\, because problem is rather about FS rather than OS\.

**Comment by p5pRT** [comment]
### From @demerphq

On 17 October 2012 18&#8203;:35\, Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> wrote&#8203;:
> \* bulk 88 via RT \<perlbug\-followup@&#8203;perl\.org> \[2012\-10\-07T14&#8203;:28&#8203;:18\]
>> I dont see a portable way of doing this if you include case insensitive
>> and not case preserving filing systems like FAT16 on DOS\.
>
> Do we still support this configuration in general?
>
> That is&#8203;:
>
> 1\. specifically\, do we work on a FAT16 file system?
>
> 2\. generally\, are there other such filesystems on which we'd like to work?

To me this comes down to being able to detect if a file system is both
case insensitive and case preserving\.

If we can detect this then we should\, in places like use/require
verify that what we have asked to load is what we got\.

If we cannot detect this then we should do nothing\.

It makes no sense to argue that because there exists a case where we
cannot detect this that we should not do so when and if we can\.

We want to make programming easier on Mac's and Windows\. Afaik both of
them in modern configuration come with file systems that are
case\-preserving but case\-insensitivity\. If we can detect that they are
in this mode we should help the user not to shoot their foot off\.

IMO even if we were to get false positives on some "weird" file
systems \(like FAT16\) and configuration this warning would be
sufficiently useful that it would not be an unreasonable burden to
make users of those configurations turn the warning off somehow\,
perhaps with a command line switch\.

Anyway\, IMO the only reason not to do this is that we have so many
false positives that it is useless everywhere\, which IMO is not
likely\.

Yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On 18 October 2012 11&#8203;:34\, Konovalov\, Vadim \(Vadim\)\*\* CTR \*\*
\<vadim\.konovalov@&#8203;alcatel\-lucent\.com> wrote&#8203;:
>> From&#8203;: demerphq \[mailto&#8203;:demerphq@&#8203;gmail\.com\]
>> On 18 October 2012 11&#8203;:05\, Konovalov\, Vadim wrote&#8203;:
>> >> From&#8203;: Ricardo Signes \[mailto&#8203;:perl\.p5p@&#8203;rjbs\.manxome\.org\]
>> >> \* bulk 88 via RT \<perlbug\-followup@&#8203;perl\.org> \[2012\-10\-07T14&#8203;:28&#8203;:18\]
>> >> > I dont see a portable way of doing this if you include case
>> >> insensitive
>> >> > and not case preserving filing systems like FAT16 on DOS\.
>> >>
>> >> Do we still support this configuration in general?
>> >>
>> >> That is&#8203;:
>> >>
>> >> 1\. specifically\, do we work on a FAT16 file system?
>> >
>> > IMO this FAT16 thing is no different compared to FAT32 and
>> even to NTFS \-
>> > these are case insensitive\, case forgiving file systems\,
>>
>> What you just said makes no sense\. This thread is about detecting
>> misloads due to case insensitive file systems\.
>
> ok\,
>
> my message was an answer to RJBS's question about FAT16\,
>
> so I pointed that this "FAT16" is exactly the same as FAT32 and NTFS\,
> WRT the problem described in ticket\,

You stated they were all case\-insensitive\. But the key point is
whether they are case\-\*preserving\*\.

I seem to recall that DOS/FAT16 was not case preserving\. Whereas I
believe that FAT32 is case preserving and know that NTFS is case
preserving\.

As far as I understand it this was the reason that Ricardo was asking
if we run under that configuration\. Which as I point out in reply to
him only matters if we did but could not detect it\.

>>
>> > Even this is not refering to 8\.3 restrictions&#8203;: FAT16 is
>> able to deal with long
>> > names\. \(so DOS perl could suffer some difficulties\, but
>> those are not
>> > related to the ticket\)
>>
>> I dont understand what this has to do with the original ticket?
>
> maybe I misunderstood the point \- but 1\) original ticket was not
> refering to DOS \(which could have 8\.3 restriction on file names\)
> and 2\) was about things like
>
> use Strict;
>
> Somehow\, DOS appeared later in discussions\, and this
> misleading\, because problem is rather about FS rather than OS\.

Yes\, this thread is about whether we should warn if we can detect that
a mis\-load has occurred\, which pretty much requires that we talk about
detecting whether a file system is both case\-preserving and
case\-insensitive\.

Yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @jandubois

On Wed\, Oct 17\, 2012 at 9&#8203;:35 AM\, Ricardo Signes
\<perl\.p5p@&#8203;rjbs\.manxome\.org> wrote&#8203;:
> \* bulk 88 via RT \<perlbug\-followup@&#8203;perl\.org> \[2012\-10\-07T14&#8203;:28&#8203;:18\]
>> I dont see a portable way of doing this if you include case insensitive
>> and not case preserving filing systems like FAT16 on DOS\.
>
> Do we still support this configuration in general?
>
> That is&#8203;:
>
> 1\. specifically\, do we work on a FAT16 file system?

FAT16 could theoretically be used on memory cards and USB sticks that
can still be used on Windows XP \(but not Windows 7\)\.  Extremely
unlikely \(because FAT16 is limited to 2GB\)\, put possible\.

> 2\. generally\, are there other such filesystems on which we'd like to work?

I haven't seen this in a long while\, but CDFS \(ISO 9660\) only supports
uppercase characters\.  I've seen a fair number of bug reports from
Windows users because the ActiveState PerlApp utility uses a
case\-sensitive algorithm internally and couldn't find MYMODULE\.PM when
they tried to "require MyModule"\.

Again\, legacy technology\, but potentially still supported\.

Cheers\,
\-Jan

**Comment by p5pRT** [comment]
### From @b2gills

On Thu\, Oct 18\, 2012 at 4&#8203;:48 AM\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
> On 18 October 2012 11&#8203;:34\, Konovalov\, Vadim \(Vadim\)\*\* CTR \*\*
> \<vadim\.konovalov@&#8203;alcatel\-lucent\.com> wrote&#8203;:
>>> From&#8203;: demerphq \[mailto&#8203;:demerphq@&#8203;gmail\.com\]
>>> On 18 October 2012 11&#8203;:05\, Konovalov\, Vadim wrote&#8203;:
>>> >> From&#8203;: Ricardo Signes \[mailto&#8203;:perl\.p5p@&#8203;rjbs\.manxome\.org\]
>>> >> \* bulk 88 via RT \<perlbug\-followup@&#8203;perl\.org> \[2012\-10\-07T14&#8203;:28&#8203;:18\]
>>> >> > I dont see a portable way of doing this if you include case
>>> >> insensitive
>>> >> > and not case preserving filing systems like FAT16 on DOS\.
>>> >>
>>> >> Do we still support this configuration in general?
>>> >>
>>> >> That is&#8203;:
>>> >>
>>> >> 1\. specifically\, do we work on a FAT16 file system?
>>> >
>>> > IMO this FAT16 thing is no different compared to FAT32 and
>>> even to NTFS \-
>>> > these are case insensitive\, case forgiving file systems\,
>>>
>>> What you just said makes no sense\. This thread is about detecting
>>> misloads due to case insensitive file systems\.
>>
>> ok\,
>>
>> my message was an answer to RJBS's question about FAT16\,
>>
>> so I pointed that this "FAT16" is exactly the same as FAT32 and NTFS\,
>> WRT the problem described in ticket\,
>
> You stated they were all case\-insensitive\. But the key point is
> whether they are case\-\*preserving\*\.
>
> I seem to recall that DOS/FAT16 was not case preserving\. Whereas I
> believe that FAT32 is case preserving and know that NTFS is case
> preserving\.

FAT32 and FAT16 and FAT12 \(floppy\)
Are all case preserving\, if you are using DOS 7\+ \(Win95\+\)
\(Win9x computers start in DOS 7\+\, and then boot into Windows\)

They support LFN filenames\. LFN is the technology that allowed Microsoft
to support filenames longer than 8\.3

If you actually took a hex editor and looked at the directory
structure you would see
two entries for every file and folder\.\( applies to FAT\(12|16|32\) \)

PROGRA~1
Program Files\.lfn

If you put another file/folder that starts with "progra" it would be stored as

PROGRA~2\.TXT
programs\.txt



---
If FAT16 wasn't case preserving\, then Windows 95 wouldn't work
correctly on it\.
I know for a fact that early Win95 computers came installed on a
FAT16 partition\. They even came with a utility to convert it to a FAT32
partition\.

This means that this discussion about FAT\(12|16|32\) and case preserving
filesystems\, is completely pointless\.



---
Because Windows was designed to be backwards compatible\, it had to be
case insensitive\. That also means it had to store the names twice\, once in 8\.3
and another in LFN\.
\( For programs which weren't made for filenames longer than 8\.3 \)

Windows can't support two names that differ only in case\.
\( I think a custom driver for FAT could potentially support it though \)
So that part of the discussion is still valid\.

>
> As far as I understand it this was the reason that Ricardo was asking
> if we run under that configuration\. Which as I point out in reply to
> him only matters if we did but could not detect it\.
>
>>>
>>> > Even this is not refering to 8\.3 restrictions&#8203;: FAT16 is
>>> able to deal with long
>>> > names\. \(so DOS perl could suffer some difficulties\, but
>>> those are not
>>> > related to the ticket\)
>>>
>>> I dont understand what this has to do with the original ticket?
>>
>> maybe I misunderstood the point \- but 1\) original ticket was not
>> refering to DOS \(which could have 8\.3 restriction on file names\)
>> and 2\) was about things like
>>
>> use Strict;
>>
>> Somehow\, DOS appeared later in discussions\, and this
>> misleading\, because problem is rather about FS rather than OS\.

Actually this problem is more about OS than FS

FS does have something to do with it\. It just isn't as limiting as you think\.

>
> Yes\, this thread is about whether we should warn if we can detect that
> a mis\-load has occurred\, which pretty much requires that we talk about
> detecting whether a file system is both case\-preserving and
> case\-insensitive\.
>

Which means we can stop talking about FAT\(12|16|32\) because they are
all case\-preserving\, if the OS supports it\.

**Comment by p5pRT** [comment]
### From @xdg

On Thu\, Oct 18\, 2012 at 1&#8203;:41 PM\, Brad Gilbert \<b2gills@&#8203;gmail\.com> wrote&#8203;:
> Actually this problem is more about OS than FS

Exactly\.

Either an OS \(that can read case\-insensitive file systems\) has an API
for determining if exact\-case $PATH exists or it doesn't&#8203;:

\(1\) it does \-\- then we have two options&#8203;:

&nbsp;  \(1a\) if exact\-case $PATH doesn't exist\, require\(\) dies

&nbsp;  \(1b\) if exact\-case $PATH doesn't exist\, but case\-insensitive $PATH
does\, require\(\) it but warn; otherwise die

\(2\) it doesn't \-\- then users run the risk of loading wrong modules and
there's nothing we can do\, so tough

I would prefer \(1a\) over \(1b\)\.

David

\-\- 
David Golden \<xdg@&#8203;xdg\.me>
Take back your inbox\! → http&#8203;://www\.bunchmail\.com/
Twitter/IRC&#8203;: @&#8203;xdg

**Comment by p5pRT** [comment]
### From @jandubois

On Thu\, Oct 18\, 2012 at 11&#8203;:11 AM\, David Golden \<xdg@&#8203;xdg\.me> wrote&#8203;:
> Either an OS \(that can read case\-insensitive file systems\) has an API
> for determining if exact\-case $PATH exists or it doesn't&#8203;:

That doesn't help you if somebody copies software from an
all\-uppercase CDROM to a case\-preserving NTFS hard disk\.

But I would be willing to just call those unsupported configurations
because the number of people benefitting from this new warning would
be a lot larger than the people inconvenienced by the change\. And they
could continue to use Perl 5\.16 if they can't upgrade their file
systems\.

However\, was there actually an efficient implementation that allowed
you to check the filename? I remember that we had this discussion many
years ago\, and that these checks could be very expensive if the module
directory was on a network filesystem\.

**Comment by p5pRT** [comment]
### From @xdg

On Thu\, Oct 18\, 2012 at 2&#8203;:25 PM\, Jan Dubois \<jand@&#8203;activestate\.com> wrote&#8203;:
>
> However\, was there actually an efficient implementation that allowed
> you to check the filename? I remember that we had this discussion many
> years ago\, and that these checks could be very expensive if the module
> directory was on a network filesystem\.

Keep in mind that I \*think\* we're only talking about require\(\) and not
open\(\) in general\.  If someone copied Perl modules from an uppercase
CDROM and wanted to require them without fixing up case\.\.\. well\.\.\. I
think we're getting into "sorry\, unsupported" territory\.

If we really think people are running Perl off a network disk and
prefer speed over precision\, then any change to check case could be
toggled at Configure time\.

I'm not terribly bothered leaving it how it is\.  If you're running a
case\-insenstive filesystem\, you've got to expect some oddities\.

\-\- 
David Golden \<xdg@&#8203;xdg\.me>
Take back your inbox\! → http&#8203;://www\.bunchmail\.com/
Twitter/IRC&#8203;: @&#8203;xdg

**Comment by p5pRT** [comment]
### From @ap

\* Brad Gilbert \<b2gills@&#8203;gmail\.com> \[2012\-10\-18 19&#8203;:45\]&#8203;:
> If FAT16 wasn't case preserving\, then Windows 95 wouldn't work
> correctly on it\.

VFAT is case\-insensitive\. Case\-preserving or not therefore cannot make
any difference to whether Windows 95 would work on it\. What would break
Win95 is if you threw away all the LFN entries on the VFAT file system
it is installed on\.

> I know for a fact that early Win95 computers came installed on a FAT16
> partition\. They even came with a utility to convert it to a FAT32
> partition\.

In light of the above\, this is a red herring\.

> This means that this discussion about FAT\(12|16|32\) and case
> preserving filesystems\, is completely pointless\.

The real question is which file system API the DOS platform support in
perl is written against – is it the old pre\-LFN DOS API or the newer one
from DOS 7? If the old API\, then the presence or absence of LFNs makes
no difference to perl and the file system will appear case\-destructive
whether or not it is\. \(It will also be limited to 8\+3\.\)

Regards\,
\-\- 
Aristotle Pagaltzis // \<http&#8203;://plasmasturm\.org/>

**Comment by p5pRT** [comment]
### From @b2gills

On Thu\, Oct 18\, 2012 at 2&#8203;:40 PM\, Aristotle Pagaltzis \<pagaltzis@&#8203;gmx\.de> wrote&#8203;:
> \* Brad Gilbert \<b2gills@&#8203;gmail\.com> \[2012\-10\-18 19&#8203;:45\]&#8203;:
>> If FAT16 wasn't case preserving\, then Windows 95 wouldn't work
>> correctly on it\.

Actually I'm not sure that the Case\-preservation is needed directly by Windows\,
just the length of the names\.

\- When Windows is installed\, many of the registry entries have the long name\.
Which means that the file\-system must support long names\.
\- The only file\-systems Win9x supported were FAT based\.
\- The only \(afaik\) system built on FAT that Windows supports is LFN\.
\- LFN is case\-preserving

All of these points were compressed into that one statement\.

>
> VFAT is case\-insensitive\. Case\-preserving or not therefore cannot make
> any difference to whether Windows 95 would work on it\. What would break
> Win95 is if you threw away all the LFN entries on the VFAT file system
> it is installed on\.

Ok the only thing Win9x really needed was the length\.
It's just the only system it supported for that\, also provided
case\-preservation\.

>
>> I know for a fact that early Win95 computers came installed on a FAT16
>> partition\. They even came with a utility to convert it to a FAT32
>> partition\.
>
> In light of the above\, this is a red herring\.
>
>> This means that this discussion about FAT\(12|16|32\) and case
>> preserving filesystems\, is completely pointless\.
>
> The real question is which file system API the DOS platform support in
> perl is written against – is it the old pre\-LFN DOS API or the newer one
> from DOS 7? If the old API\, then the presence or absence of LFNs makes
> no difference to perl and the file system will appear case\-destructive
> whether or not it is\. \(It will also be limited to 8\+3\.\)
>

I'm not sure that perl still compiles for DOS\.
The newest reference I could come up with is for Perl 5\.8\.8
\(DJGPP itself doesn't appear to have any activity since 2003\)

According to README\.dos
http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/HEAD&#8203;:/README\.dos
it doesn't use any special API added to DOS7

I'm not sure how the API changed in DOS7\.
DOS7 itself understands the LFN extension; I'm just not sure how it allows
a programmer to access this extra information\.

perl does utilize Win9x functions\, if you set up the environment correctly
before unpacking the archive\. \(according to README\.dos\)
\( provided by DJGPP \)



---
I would like to point out that Win9x support has been removed\.

I think we can remove DOS support as well\, without anyone complaining loudly\.
